Change the email authenticator to use a database table instead of ses…

EreMaijala opened the pull request:
…sion storage.

I tested that the pgsql migration works, but I haven't tried actually using it.

TODO
- [x] Test with PostgreSQL
- [x] Update documentation to note need to run expiration utility
- [x] Update database changelog

EreMaijala committed:
Change email authenticator to use database instead of session storage.

demiankatz commented:
Thanks, @EreMaijala. I'll do some end-to-end testing with this using PostgreSQL next week when I have a bit more time. In the meantime, one question: is this methodology less secure than the old approach? That is, the old way, when a user got a link, it would only actually work if they had an existing browser session that the link would resume. If somebody intercepted the link, it would be useless, because they wouldn't have the associated session ID. This method makes the link more robust (I understand that allowing this to work across multiple browsers was a requirement), but it also seems to greatly increase the danger of link interception and abuse. Maybe I missed a double-check of some sort when I was reviewing the code... but if I didn't, I wonder if we should add a configuration setting to make it possible to force validation of the stored session ID against the user's current session ID. That would allow users to choose between more secure vs. more flexible.

demiankatz commented:
...another possibility would be to add some kind of multi-factor user validation; i.e. save some other fingerprints in the database (e.g. IP address) and require at least one thing to match to accept the link. This would still be vulnerable to spoofing, etc., but it might be better than nothing.

EreMaijala commented:
@demiankatz I think you concern is valid, and this makes the mechanism less robust. Adding at least an IP address check would probably make sense. I'll consider other options too, but I think other fingerprinting methods may have the same problem of using different browsers or private windows messing them up.

EreMaijala committed:
Add IP address check to EmailAuthenticator.

EreMaijala commented:
I've added IP address checking. I believe this to be sufficient since the IP address is not shared in the email or its envelope. It's still possible that a shared IP address could allow an attacker access, but at least that would require being in the same subnet as the user and accessing the login link before the user does.

To make things even more secure we could turn the whole thing around and just send a one-time code in the email for the user to enter in the existing session. This would come with a slightly higher cognitive load, and I think in that case we should consider doing the same for the email address verification mechanism.

EreMaijala commented:
A couple of comments about the latest commit: the Request parameter to EmailAuthenticator's constructor wasn't used (oops!) so I replaced it with RemoteAddress. I added a factory that's really a copy of the ServerUrlFactory class, so I kept the copyright intact.

demiankatz commented:
Thanks, @EreMaijala, just one more thought: our IP address check is now quite strict, and some users may be subject to frequently changing IP addresses (I see this a lot with mobile devices, for example, if people are working while in motion). What if we approve the user if either the IP address OR the session ID matches? That will cover change of browser and change of IP address, only failing in the (hopefully) rare case that a user's IP address changes at the exact same time they decide to switch to a different browser.

EreMaijala committed:
Allow login also when session id matches even if IP address doesn't.

EreMaijala commented:
@demiankatz Good point. Now the session id is checked as well, and either one matching will suffice.

demiankatz commented on module/VuFind/src/VuFind/Http/PhpEnvironment/RemoteAddressFactory.php:
```diff
@@ -0,0 +1,72 @@
+        }
+        $cfg = $container->get(\VuFind\Config\PluginManager::class)->get('config');
+        $object = new $requestedName();
+        if ($cfg->Site->reverse_proxy ?? false) {
```
This appears to be an undocumented setting. Can you add some notes to config.ini? Is this something that actually belongs here, or should this be committed separately?

EreMaijala replied:
Not something I came up with, but I'm pretty proud of having taken it into account. ;) See:

https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/View/Helper/Root/ServerUrlFactory.php#L67
https://github.com/vufind-org/vufind/blob/master/config/vufind/config.ini#L29

demiankatz replied:
I don't know what I was looking at, but clearly it was the wrong thing. This was added in #1429, and as you say, it's already documented. Sorry about the confusion (and good work accounting for it). :-)

demiankatz committed:
Merge branch 'master' into email-auth-hash

demiankatz commented:
I've just merged the language file changes from here into master -- it's time for me to start requesting translations for the 6.1 release, and I want to be sure everything is up to date. I'm confident that the rest of this will follow very soon.

demiankatz commented:
@EreMaijala, I was hoping to test and merge this this morning, but unfortunately, I now can't seem to get it to work at all. I tested with both PostgreSQL and MySQL, and in both cases, I don't think the email authentication is working. I simply turned on the email option in the [Authentication] section, and this does cause the login prompt to turn into an email-only box. However, no matter what I enter into the box, I get the expected message indicating that an email has been sent, but I never receive the email. I'm concerned that the software isn't even trying to send the email, because if I set up invalid SMTP credentials in the [Mail] section of config.ini, it still appears to succeed even though it can't possibly. Additionally, if I enter invalid email addresses (like just "foo"), it also appears to succeed. Are you seeing the same behavior on your end, or am I just doing something wrong?

EreMaijala commented:
@demiankatz Use the Demo driver and set it to use email authentication. It won't work standalone. See the original PR https://github.com/vufind-org/vufind/pull/1409#issuecomment-523376164 for more information.

demiankatz commented:
@EreMaijala, I'm confused; I thought it started out that way, and then we worked through making it work as a generic authentication option. I must have formed some false memories somewhere along the way! At very least, I think we need to figure out a way to throw an exception when things are not configured correctly, because otherwise this is a very confusing pitfall. I'm out of time for today, but I'll look at this more closely next week.

EreMaijala commented:
@demiankatz Ah, yes, sorry. I lost track of my thoughts. I'll recheck.

EreMaijala committed:
Fix Email authentication.

EreMaijala commented:
@demiankatz Should be fixed now. I'm not sure if it ever actually worked, but now it's at least possible.

demiankatz committed:
Merge branch 'master' into email-auth-hash

demiankatz committed:
Case insensitive method validation.

demiankatz commented:
@EreMaijala, I'm still having problems.

I realized that my problems on Friday were related to the fact that Email authentication cannot create new accounts, it can only log into existing ones. I had forgotten that detail, and because we always report success as a security measure, even when email addresses don't exist in the database, it left me confused.

With that out of the way, I set this up correctly with ChoiceAuth offering Database and Email options. I then created an account with the Database method, and tried to log in using the Email option. Now I am successfully receiving email messages... but the login links do not work.

First, the login links took me to a fatal error screen, related to 'email' not being a valid option for ChoiceAuth (my config.ini said 'Email' instead). I made the comparison case insensitive in bee7b8d to work around this, but I'm not sure if that's the best approach....

And so now I'm stuck with the links just giving me the message "We cannot log you in at this time. Please try again later."

I'm out of time for today; tomorrow, I may have time to put some debug messages into the code to figure out exactly which condition is failing... but if you have time to look at this again before I do, I'd be interested to know if you're getting better results with the Database/Email/ChoiceAuth setup.

EreMaijala committed:
Add proxying of needsCsrfCheck method to ChoiceAuth.

EreMaijala commented:
@demiankatz Right, this was probably never tested with ChoiceAuth. It was missing proxying of the needsCsrfCheck method, so Email auth never got the chance to say it's not needed. Should be fixed now. Thanks for the case-insensitivity fix!

demiankatz committed:
Merge branch 'master' into email-auth-hash

demiankatz committed:
Add comment.

demiankatz commented:
Thanks, @EreMaijala, now it's working as expected. I really thought we had this working before, but I must have neglected to test something fully. In any case, I'm merging now, so I think we're finally in good shape. Thanks for all the work on this!
